### DISCUSSION TITLE
querynode oom

### DISCUSSION THREAD

**Discussion by luckercs** [milvus-io/milvus#27157] [Q&A and General discussion] [answered]
在我们的集群中，扩展了20多个querynode，单个querynode的resource limit为16gi，当创建一个collection并创建500个partition，每个分区导入30000条数据时，发现刚开始只有1个querynode的内存在逐步扩增，当该pod的内存近满负载时，约等了5分钟之后，所有querynode的内存才开始逐步均衡，但是此时查询这个collection的num_entities数，发现和预期比较会缺失一部分数据，查看querynode，发现有报错如下：

> `xxxx [WARN] [querynode/impl_utils.go:55] ["insufficient memory when shard cluster load segments"] [traceID=70d4c194f4cbf82a] [shard=milvus-my-release-rootcoord-
dml_0_444320398254277960v0] [segmentIDs="[444320398270730685]"] [error="InsufficientMemoryToLoad, reason:InsufficientMemoryToLoad, load segment failed, OOM if load, collectionID = 44432039
8254277960, MaxSegmentMemUsage = 46 MB, concurrency = 1, usedMemAfterLoad = 15623 MB, currentAvailableMemCount = 806 MB, currentTotalMem = 16384 MB, thresholdFactor = 0.900000"]
[2023/09/17 12:17:38.423 +00:00] [INFO] [querynode/segment_loader.go:879] ["predict memory and disk usage while loading (in MiB)"] [collectionID=444320398254277960] [concurrency=1] [maxSeg
mentSize=46] [memUsage=15577] [freeMemory=806] [totalMemory=16384] [predictMemUsage=15623] [predictPeakMemUsage=15670] [diskUsage=0] [predictDiskUsage=0] [totalDisk=3755428]
`

> `xxxx  [ERROR] [querynode/segment_loader.go:153] ["load failed, OOM if loaded"] [collectionID=444320398254277960] [segmentType=Sealed] ["loadSegmentRequest msgID"
=1307] [error="InsufficientMemoryToLoad, load segment failed, OOM if load, collectionID = 444320398254277960, MaxSegmentMemUsage = 46 MB, concurrency = 1, usedMemAfterLoad = 15623 MB, curr
entAvailableMemCount = 806 MB, currentTotalMem = 16384 MB, thresholdFactor = 0.900000"] [stack="github.com/milvus-io/milvus/internal/querynode.(*segmentLoader).LoadSegment\n\t/go/src/githu
b.com/milvus-io/milvus/internal/querynode/segment_loader.go:153\ngithub.com/milvus-io/milvus/internal/querynode.(*loadSegmentsTask).Execute\n\t/go/src/github.com/milvus-io/milvus/internal/
querynode/load_segment_task.go:81\ngithub.com/milvus-io/milvus/internal/querynode.(*taskScheduler).processTask\n\t/go/src/github.com/milvus-io/milvus/internal/querynode/task_scheduler.go:1
19\ngithub.com/milvus-io/milvus/internal/querynode.(*taskScheduler).taskLoop\n\t/go/src/github.com/milvus-io/milvus/internal/querynode/task_scheduler.go:141"]`

在大量插入数据时，如何避免单个querynode oom发生，或者有哪些参数可以调整，比如在开始插入数据时，设置多个querynode的并发？

**Comment by xiaofan-luan** [comment]
> 在我们的集群中，扩展了20多个querynode，单个querynode的resource limit为16gi，当创建一个collection并创建500个partition，每个分区导入30000条数据时，发现刚开始只有1个querynode的内存在逐步扩增，当该pod的内存近满负载时，约等了5分钟之后，所有querynode的内存才开始逐步均衡，但是此时查询这个collection的num_entities数，发现和预期比较会缺失一部分数据，查看querynode，发现有报错如下：
> 
> > `xxxx [WARN] [querynode/impl_utils.go:55] ["insufficient memory when shard cluster load segments"] [traceID=70d4c194f4cbf82a] [shard=milvus-my-release-rootcoord- dml_0_444320398254277960v0] [segmentIDs="[444320398270730685]"] [error="InsufficientMemoryToLoad, reason:InsufficientMemoryToLoad, load segment failed, OOM if load, collectionID = 44432039 8254277960, MaxSegmentMemUsage = 46 MB, concurrency = 1, usedMemAfterLoad = 15623 MB, currentAvailableMemCount = 806 MB, currentTotalMem = 16384 MB, thresholdFactor = 0.900000"] [2023/09/17 12:17:38.423 +00:00] [INFO] [querynode/segment_loader.go:879] ["predict memory and disk usage while loading (in MiB)"] [collectionID=444320398254277960] [concurrency=1] [maxSeg mentSize=46] [memUsage=15577] [freeMemory=806] [totalMemory=16384] [predictMemUsage=15623] [predictPeakMemUsage=15670] [diskUsage=0] [predictDiskUsage=0] [totalDisk=3755428] `
> 
> > `xxxx  [ERROR] [querynode/segment_loader.go:153] ["load failed, OOM if loaded"] [collectionID=444320398254277960] [segmentType=Sealed] ["loadSegmentRequest msgID" =1307] [error="InsufficientMemoryToLoad, load segment failed, OOM if load, collectionID = 444320398254277960, MaxSegmentMemUsage = 46 MB, concurrency = 1, usedMemAfterLoad = 15623 MB, curr entAvailableMemCount = 806 MB, currentTotalMem = 16384 MB, thresholdFactor = 0.900000"] [stack="github.com/milvus-io/milvus/internal/querynode.(*segmentLoader).LoadSegment\n\t/go/src/githu b.com/milvus-io/milvus/internal/querynode/segment_loader.go:153\ngithub.com/milvus-io/milvus/internal/querynode.(*loadSegmentsTask).Execute\n\t/go/src/github.com/milvus-io/milvus/internal/ querynode/load_segment_task.go:81\ngithub.com/milvus-io/milvus/internal/querynode.(*taskScheduler).processTask\n\t/go/src/github.com/milvus-io/milvus/internal/querynode/task_scheduler.go:1 19\ngithub.com/milvus-io/milvus/internal/querynode.(*taskScheduler).taskLoop\n\t/go/src/github.com/milvus-io/milvus/internal/querynode/task_scheduler.go:141"]`
> 
> 在大量插入数据时，如何避免单个querynode oom发生，或者有哪些参数可以调整，比如在开始插入数据时，设置多个querynode的并发？

如果你的partition非常均匀，每个partition可能都达不到flush一个segment的阈值（128MB），因此可能会出现导致数据全部处于growing状态，全部都在delegator所处的节点上。

以500个partition为例子，一台querynode所需要的内存大小可能是 128MB * 500 = 62.5G, 显然超过你的内存大小。

要解决这个问题
1. 看看能否使用更少的partition
2. 调小 sealProportion, 默认是0.23，一个segment的flush大小就是512MB * 0.23， 你可以调整到0.08，这样一个segment只需要40.96MB内存,500个只需要20多G
3. 这么多querynode，建议每个querynode调整到16c64G或者16c32G.
4. 是否能够一个partition一个partition的写入，而不是所有partition均匀写入
5. 关注索引构建情况，保证索引能够及时替换
6. 如果你有本地磁盘，可以尝试开启mmap，把写入数据mmap到本地磁盘上

**Reply by luckercs** [reply]
我们将sealProportion参数调小后，同时减少了querynode数，发现针对单个collection，在逐个分区插入数据并加载数据的时候，各个querynode的内存占用依然不均衡，其中一个querynode的pod内存占用仍比较明显，与其他各个querynode之间依然非常不均衡。占用内存最大的querynode将限制了插入数据的最大量？这种该如何解决呢？

**Comment by luckercs** [comment]
占用内存最大的querynode将限制了插入数据的最大量。如果需要插入并加载上亿的数据量来供查询，需要对单机querynode的内存量是否有要求？似乎querynode扩展多个replicas副本并不能解决问题 @xiaofan-luan

**Reply by luckercs** [reply]
通过查看querynode的指标监控，发现各个querynode加载的entity不均衡，这里应该是各个querynode之间的内存均衡没生效或者均衡效果不明显，我们将milvus镜像版本升级到了最新的版本2.3，配置以及测试条件均不变，再次查看监控，发现各个querynode均衡生效了，各个pod的内存使用也相对比较均衡，问题得到解决，应该是在后续升级过程中给解决了，再次感谢 @xiaofan-luan

**Reply by yhmo** [reply, marked as answer]
建表的时候有个参数shards_num，默认为1
插入数据的时候数据先写入pulsar，然后由datanode/querynode去订阅，如果某个表的shards_num=1，那么就会有某个querynode创建一条"数据管道"从pulsar拉取数据，把数据累积到内存里(前提是你调用过collection.load())，这个querynode我们称之为delegator（之前叫做shard leader）
某个datanode也会做相同的事情，只不过datanode是把拉取到的数据凑成一个分片之后持久化到存储系统，分片大小默认512MB。
我们通常说的segment，在存储系统里的文件是它的“真身”，在querynode内存里是它的"影子"，目前版本里的规则是: 在分片的索引没建好之前，它的影子总是先放在delegator的内存里，当它的索引建好之后，才会有另一个querynode把它的索引加载进自己的内存，而原先在delegator内存里的那份就释放掉。因此，你会看到密集插入的情况下，delegator的内存占用量远大于其他querynode，随着分片建好索引，delegator的内存才得到缓解。

通常，建立底库的时候，要大量地插入数据，这时候如果你没有查询的需求，可以不调用collection.load()，这样就不会有querynode消耗内存。等你把一亿条的底库建立好了，再去调用collecton.load()，这样就可以避免你的问题。如果非要在建立底库的时候去查询，那基本也没有完美的解决方法，把shards_num改为2或4可以缓解(这样就会有多个querynode共同承担压力)，不过shards_num改大了也会把数据切的更碎，对查询性能不太友好。

**Reply by luckercs** [reply]
@yhmo Thank you for explaining the Milvus architecture. I have learned a lot from it！

**Comment by yhmo** [comment]
PS: 在社区里请尽量使用英文交流，因为其他地区的用户也许有相同的问题，他们搜索时是没法从中文搜到问题和答案的。
